src/os/bluestore/BlueStore.cc

```diff
@@ -13186,6 +13188,115 @@ void BlueStore::_do_write_small(
   return;
 }
 
+bool BlueStore::BigDeferredWriteContext::can_defer(
+    BlueStore::extent_map_t::iterator ep,
+    uint64_t prefer_deferred_size,
+    uint64_t block_size,
+    uint64_t offset,
+    uint64_t l)
+{
+  bool res = false;
+  auto& blob = ep->blob->get_blob();
+  if (offset >= ep->blob_start() &&
+    blob.is_mutable()) {
+    off = offset;
+    b_off = offset - ep->blob_start();
+    uint64_t chunk_size = blob.get_chunk_size(block_size);
+    uint64_t ondisk = blob.get_ondisk_length();
+    used = std::min(l, ondisk - b_off);
+
+    // will read some data to fill out the chunk?
+    head_read = p2phase<uint64_t>(b_off, chunk_size);
+    tail_read = p2nphase<uint64_t>(b_off + used, chunk_size);
+    b_off -= head_read;
```

Comment:
Is it possible to get head_read != 0 or tail_read != 0 ?
I think that such cases will be caught by do_write_small.

Reply:
It's possible to have chunk size > alloc unit sometimes. E.g. when expected_write_size is provided, see _choose_write_options(...)
Hence we one can get head/tail at checksum chunk boundaries rather than at allocation unit ones. But do_write_small/do_write_big are distinguished on the latter only.

Reply:
@ifed01 
I was not sure on how _do_write_small and _do_write_big_deferred interact for different chunk_sizes, so I tested it.
Do you think it is useful to grab it here?
https://github.com/ceph/ceph/commit/7762a8232d902095c203a91b873614ab583e4178

Reply:
Absolutely! More tests are always better :) Will add, thanks!